src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,168 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = tail = new Node(null, value, null);
+        } else {
+            Node notTheTailAnyMore = tail;
+            tail = new Node(notTheTailAnyMore, value, null);
+            notTheTailAnyMore.next = tail;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            if (index == size) {
+                add(value);
+            } else if (index == 0) {
+                Node notTheHeadAnyMore = head;
+                head = new Node(null, value, notTheHeadAnyMore);
+                notTheHeadAnyMore.prev = head;
+                size++;
+            } else {
+                int currentIndex = 0;
+                Node currentNode = head;
+                while (currentIndex != index) {
+                    currentNode = currentNode.next;
+                    currentIndex++;
+                }
+                Node newNode = new Node(currentNode.prev, value, currentNode);
+                currentNode.prev.next = newNode;
+                currentNode.prev = newNode;
+
+                size++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new IndexOutOfBoundsException("given index does not exist");
+        } else {
+            int currentIndex = 0;
```

Comment:
Don't iterate your nodes from the head to the tail or vice versa. Pay attention to the index you receive as an input, you can iterate only the first or second half of the list, depending on the index value. That will boost your lists' performance.